lib/src/main/scala/spinal/lib/bus/regif/Document/CHeaderGenerator.scala

```diff
@@ -52,54 +54,165 @@ final case class CHeaderGenerator(
+                def mask = BigInt((1 << fd.getSection().size) - 1) << lsb
+                val _tab = " " * (tabn - fd.getName().size)
+                fd.getAccessType() match {
+                    case `NA` |`RO` |`ROV`                       => ""
+                    case `W1S`|`W1C`|`W1T`|`W1P`|`W1CRS`|`W1SRC` => ""
+                    case `W0S`|`W0C`|`W0T`|`W0P`|`W0CRS`|`W0SRC` => ""
+                    case _ => {
+                        s"""#define ${pre}_${fd.getName().toUpperCase()}_SHIFT ${_tab}${lsb}
+                           |#define ${pre}_${fd.getName().toUpperCase()}_MASK  ${_tab}0x${mask.hexString(32)} //${fd.getAccessType()}, ${fd.getWidth()} bit""".stripMargin
+                    }
+                }
+            }
+        }
+
+        def body() = {
+            val maxnamelen = regs.map(_.getName().size).max
+            val maxshiftlen = regs.map(t => t.getName().size + t.fdnamelens).max
+            def header: String = headers.mkString("\n * ")
             s"""|/*
-                | * Bus Interface Header - AUTOGENERATED
+                | * Reg Interface C-Header [AUTOGENERATE by SpinalHDL]
+                | * ${header}
                 | */
                 |
                 |#ifndef ${guardName}
                 |#define ${guardName}
                 |
-                |""".stripMargin)
-        
-        for(reg <- regs) {
-            pw.write(s"#define ${reg.getName.toUpperCase()} ")
-            pw.write(" " * (regLength - reg.getName.length))
-            pw.write("0x")
-            pw.print(reg.getAddr.formatted(s"%0${4}x"))
-            pw.println()
+                |${regs.map(_.define(maxnamelen, maxshiftlen)).mkString("\n")}
+                |
+                |${regs.map(_.union).mkString("\n\n")}
+                |
+                |#endif /* ${guardName} */
+                |""".stripMargin
         }
-        pw.println()
-
-        for(t <- types) {
-            val naName = "reserved_"
-            var i = -1
-            val len = math.max(naName.length + 1, t.fields.map(_.getName.length()).fold(0)(math.max(_, _)))
-
-            pw.println("typedef union {")
-
-            pw.println(s"\t$regType val;")
-
-            pw.println("\tstruct {")
-            t.fields.foreach( f => {
-                val name = f.getAccessType match {
-                    case AccessType.NA => {
-                        i += 1
-                        s"reserved_${i}"
-                    }
-                    case default => f.getName
-                }
-                pw.write(s"\t\t$regType ${name} ")
-                pw.write(" " * (len - name.length))                
-                pw.println(s": ${f.getWidth.formatted("%2d")};")
-            })
+        pw.write(body())
+        pw.close()
 
-            pw.println("\t} reg;")
-            pw.println(s"} ${t.name.toLowerCase()}_t;\n")
+        def unionregs = {
+            //todo: all regbank as union struct for mem copy usage
```

Comment:
Is it for this PR? Or should we open an issue for that and turn the todo into a fixme?